lightning/src/util/persist.rs

```diff
@@ -249,10 +259,1015 @@ where
+	///
+	/// # Parallelizing
+	/// Loading a large number of monitors will be faster if done in parallel. You can use this
+	/// function to accomplish this. Take care to limit the number of parallel readers.
+	pub fn read_channel_monitor_with_updates<B: Deref, F: Deref + Clone>(
+		&self, broadcaster: &B, fee_estimator: &F, monitor_key: String,
+	) -> io::Result<(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let monitor_name = MonitorName::new(monitor_key)?;
+		// ...parse the monitor
+		let (block_hash, monitor) = self.read_monitor(&monitor_name)?;
+		// ...parse and apply the updates with an id higher than the monitor.
+		let mut current_update_id = monitor.get_latest_update_id();
+		loop {
+			current_update_id = match current_update_id.checked_add(1) {
+				Some(next_update_id) => next_update_id,
+				None => break,
+			};
+			let update_name = UpdateName::from(current_update_id);
+			let update = match self.read_monitor_update(&monitor_name, &update_name) {
+				Ok(update) => update,
+				Err(err) if err.kind() == io::ErrorKind::NotFound => {
+					// We can't find any more updates, so we are done.
+					break;
+				}
+				Err(err) => return Err(err),
+			};
+
+			monitor
+				.update_monitor(&update, broadcaster, fee_estimator.clone(), &self.logger)
+				.map_err(|e| {
+					log_error!(
+						self.logger,
+						"Monitor update failed. monitor: {} update: {} reason: {:?}",
+						monitor_name.as_str(),
+						update_name.as_str(),
+						e
+					);
+					io::Error::new(io::ErrorKind::Other, "Monitor update failed")
+				})?;
+		}
+		Ok((block_hash, monitor))
+	}
+
+	/// Read a channel monitor.
+	fn read_monitor(
+		&self, monitor_name: &MonitorName,
+	) -> io::Result<(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)> {
+		let outpoint: OutPoint = monitor_name.try_into()?;
+		let mut monitor_cursor = io::Cursor::new(self.kv_store.read(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			monitor_name.as_str(),
+		)?);
+		// Discard the sentinel bytes if found.
+		if monitor_cursor.get_ref().starts_with(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL) {
+			monitor_cursor.set_position(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() as u64);
+		}
+		match <(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>::read(
+			&mut monitor_cursor,
+			(&*self.entropy_source, &*self.signer_provider),
+		) {
+			Ok((blockhash, channel_monitor)) => {
+				if channel_monitor.get_funding_txo().0.txid != outpoint.txid
+					|| channel_monitor.get_funding_txo().0.index != outpoint.index
+				{
+					log_error!(
+						self.logger,
+						"ChannelMonitor {} was stored under the wrong key!",
+						monitor_name.as_str()
+					);
+					return Err(io::Error::new(
```

Comment:
nit: consider making this an if-else so you don't need an explicit return here

Reply:
done